src/Entity/GeneratedBook.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace App\Entity;
+
+use App\Book;
+use DateTime;
+use DateTimeZone;
+// mediawiki-codesniffer apparently can't search annotations for unused classes.
+// phpcs:ignore MediaWiki.Classes.UnusedUseStatement.UnusedUse
+use Doctrine\ORM\Mapping as ORM;
+
+/**
+ * The GeneratedBook is the log of exports.
+ * @ORM\Entity(repositoryClass="App\Repository\GeneratedBookRepository")
+ * @ORM\Table(
+ *     name="books_generated",
+ *     indexes={
+ * @ORM\Index(name="time", columns={"time"}),
+ * @ORM\Index(name="lang", columns={"lang"})
+ *     }
+ * )
+ */
+class GeneratedBook {
+
+	/**
+	 * @ORM\Id()
+	 * @ORM\GeneratedValue()
+	 * @ORM\Column(type="integer")
+	 * @var int
+	 */
+	private $id;
+
+	/**
+	 * @ORM\Column(type="datetime")
+	 * @var DateTime Time of the export.
+	 */
+	private $time;
+
+	/**
+	 * @ORM\Column(type="string", length=10)
+	 * @var string Language code.
+	 */
+	private $lang;
+
+	/**
+	 * @ORM\Column(type="string", length=255)
+	 * @var string Title of work.
+	 */
+	private $title;
+
+	/**
+	 * @ORM\Column(type="string", length=10)
+	 * @var string Format such as 'epub', 'pdf', etc.
+	 */
+	private $format;
+
+	/**
+	 * GeneratedBook constructor.
+	 * @param Book $book
+	 * @param string $format
+	 */
+	public function __construct( Book $book, string $format ) {
+		$this->time = new DateTime( 'now', new DateTimeZone( 'UTC' ) );
+		$this->lang = $book->lang;
+		$this->setTitle( $book->title );
+		$this->format = $format;
+	}
+
+	/**
+	 * @return int|null
+	 */
+	public function getId(): ?int {
+		return $this->id;
+	}
+
+	/**
+	 * @return DateTime
+	 */
+	public function getTime(): DateTime {
+		return $this->time;
+	}
+
+	/**
+	 * @param DateTime $time
+	 */
+	public function setTime( DateTime $time ): void {
+		$this->time = $time;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getLang(): string {
+		return $this->lang;
+	}
+
+	/**
+	 * @param string $lang
+	 */
+	public function setLang( string $lang ): void {
+		$this->lang = $lang;
+	}
+
+	/**
+	 * @return string
+	 */
+	public function getTitle(): string {
+		return $this->title;
+	}
+
+	/**
+	 * @param string $title
+	 */
+	public function setTitle( string $title ): void {
+		$this->title = htmlspecialchars_decode( $title );
```

Comment:
I think `htmlspecialchars_decode` doesn't belong here, unless it's made obvious that the provided $title is expected to be encoded (e.g. by a different method name). The entity should be usable from wherever.

Not that there are many books like "The Life of %20" but who knows... :)

Reply:
This came from https://github.com/wsexport/tool/blob/151eec557a9dcc5198a4a0652a12db5edea188bb/src/CreationLog.php#L60

I didn't want to break anything by removing that. From https://phabricator.wikimedia.org/T270495 it sounds like the titles are sometimes encoded?

Reply:
Ah right! Makes sense. Let's keep it.